### ISSUE TITLE
Unhelpful error messages for trait with unbound lifetime

### ISSUE THREAD

**Issue by willir** [rust-lang/rust#72106]
I was trying to implement [FromSql](https://docs.rs/rusqlite/0.13.0/rusqlite/types/trait.FromSql.html) of `rusqlite` for my reference type (with lifetime). And I faced an unclear error message, later I was explained that the trait does not allow the return type to borrow from the value argument since it doesn't specify lifetime parameter explicitly.

However, I thought that it worth filing a bug report about unhelpful error message anyway.

Here is a simplified code:

```rust
use std::error::Error;
use std::fmt;

#[derive(Copy, Clone, Debug, PartialEq)]
pub enum ValueRef<'a> {
    Null,
    Integer(i64),
    Real(f64),
    Text(&'a [u8]),
    Blob(&'a [u8]),
}

impl<'a> ValueRef<'a> {
    pub fn as_str(&self) -> FromSqlResult<&'a str> {
        match *self {
            ValueRef::Text(t) => {
                std::str::from_utf8(t).map_err(|_| FromSqlError::InvalidType)
            }
            _ => Err(FromSqlError::InvalidType),
        }
    }
}

#[derive(Debug)]
#[non_exhaustive]
pub enum FromSqlError {
    InvalidType
}

impl fmt::Display for FromSqlError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "InvalidType")
    }
}

impl Error for FromSqlError {}

pub type FromSqlResult<T> = Result<T, FromSqlError>;

pub trait FromSql: Sized {
    fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self>;
}

impl FromSql for &str {
    fn column_result(value: ValueRef<'_>) -> FromSqlResult<&str> {
        value.as_str()
    }
}

pub fn main() {
    println!("{}", "Hello World");
}
```

The error message I'm getting is:

```rust
error: `impl` item signature doesn't match `trait` item signature
  --> src/main.rs:45:5
   |
41 |     fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self>;
   |     ------------------------------------------------------------- expected fn(ValueRef<'_>) -> std::result::Result<&str, FromSqlError>
...
45 |     fn column_result(value: ValueRef<'_>) -> FromSqlResult<&str> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ found fn(ValueRef<'_>) -> std::result::Result<&str, FromSqlError>
   |
   = note: expected `fn(ValueRef<'_>) -> std::result::Result<&str, FromSqlError>`
              found `fn(ValueRef<'_>) -> std::result::Result<&str, FromSqlError>`

error: aborting due to previous error

error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

So basically it says that I have a wrong signature, even though the *expected* and *found* signatures are the same.

I'm not sure what would be the right error message to raise here since I don't know lifetime rules quite well.

Here is, just in case, a link to [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5a524f07d85f03edaec061f8610b2614)

The rust version I'm using is: `1.44.0-nightly`, but it also reproduces on `1.43.1`.

**Comment by lcnr** [comment]
Minimized:
```rust
pub trait Foo: Sized {
    fn foo(value: &str) -> Self;
}

impl Foo for &str {
    fn foo(value: &str) -> &str {
        value
    }
}
```
Error message:
```
error: `impl` item signature doesn't match `trait` item signature
 --> src/lib.rs:6:5
  |
2 |     fn foo(value: &str) -> Self;
  |     ---------------------------- expected fn(&str) -> &str
...
6 |     fn foo(value: &str) -> &str {
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ found fn(&str) -> &str
  |
  = note: expected `fn(&str) -> &str`
             found `fn(&str) -> &str`
```

**Comment by estebank** [comment]
Duplicate of #66406. CC https://github.com/rust-lang/rust/pull/67460.

**Comment by estebank** [comment]
Update, not a duplicate. The output with #67460 is:

<img width="1004" alt="" src="https://user-images.githubusercontent.com/1606434/83165486-6e215800-a0c2-11ea-99ec-98fbb4065ec8.png">

```
error: `impl` item signature doesn't match `trait` item signature
  --> file8.rs:45:5
   |
41 |     fn column_result(value: ValueRef<'_>) -> FromSqlResult<Self>;
   |     ------------------------------------------------------------- expected `fn(ValueRef<'_>) -> std::result::Result<&str, FromSqlError>`
...
45 |     fn column_result(value: ValueRef<'_>) -> FromSqlResult<&str> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ found `fn(ValueRef<'_>) -> std::result::Result<&str, FromSqlError>`
   |
   = note: expected `fn(ValueRef<'_>) -> std::result::Result<&str, _>`
              found `fn(ValueRef<'_>) -> std::result::Result<&str, _>`
   = help: the lifetime requirements from the `impl` do not correspond to the requirements in the `trait`
   = help: verify the lifetime relationships in the `trait` and `impl` between the `self` argument, the other inputs and its output
```

We need to account for `Self` (as opposed to type parameters).
